firedrake/mesh.py

```diff
@@ -949,6 +1001,12 @@ def create_cell_node_list(self, global_numbering, fiat_element):
                                      self.cell_closure,
                                      fiat_utils.FlattenedElement(fiat_element))
 
+    def _order_data_by_cell_index(self, column_list, cell_data):
+        cell_list = []
+        for col in column_list:
+            cell_list += range(col, col + (self.layers - 1))
+        return cell_data[cell_list]
```

Comment:
This does O(N) work in pure python.  I guess if it turns out slow, we move it to cython or something.